llvm/lib/TableGen/TGLexer.cpp

```diff
@@ -54,9 +74,17 @@ TGLexer::TGLexer(SourceMgr &SM, ArrayRef<std::string> Macros) : SrcMgr(SM) {
   PrepIncludeStack.push_back(
       std::make_unique<std::vector<PreprocessorControlDesc>>());
 
-  // Put all macros defined in the command line into the DefinedMacros set.
-  for (const std::string &MacroName : Macros)
+  // Add all macros defined on the command line to the DefinedMacros set.
+  // Check invalid macro names and print fatal error if we find one.
+  for (const std::string &MacroName : Macros) {
+    const char *Begin = MacroName.c_str();
+    const char *End = lexMacroName(Begin);
```

Comment:
This assumes MacroName is null-terminated. If it really is, this should be asserted before calling the function.

Reply:
It's not exactly assuming its null terminated, just that we wil hit the end of macro name condition before we run out of characters in the supplied str. For the command line, I suspect the strings are null terminated, but the use in `prepLexMacroName` is likely not, or atleast not obviously looking at the code. So not sure what assert to add here exactly.

Reply:
It appears that std::string is guaranteed to be null terminated since C++11, so this is *probably* safe. I'm not sure if address sanitizer is happy with this.

Reply:
I don't think that is true. For example, the std::string that is passed into the Macro is constructed using StringRef::str() in CommandLine.h here:

```
  // Return true on error.
  bool parse(Option &, StringRef, StringRef Arg, std::string &Value) {
    Value = Arg.str();
    return false;
  }
```

so I don't think we can assume its null terminated. Which means that the common function I added will need to accept a StringRef, and the call from prepLexMacroName() has to pass in the appropriate length (of the string if it knows, or std::string::npos to retain the current behavior. Let me make that change, so that code functionally identical to the existing one.

Reply:
https://en.cppreference.com/w/cpp/string/basic_string/data
> The returned array is null-terminated, that is, data() and [c_str()](https://en.cppreference.com/w/cpp/string/basic_string/c_str) perform the same function.
> If [empty()](https://en.cppreference.com/w/cpp/string/basic_string/empty) returns true, the pointer points to a single null character.
> (since C++11)

Reply:
Thanks, that's good to know. With the change to use a StringRef, we don't need to worry about it now.